crates/subspace-service/src/sync_from_dsn/fast_sync.rs

```diff
@@ -0,0 +1,587 @@
+        + LockImportRun<Block, B>
+        + Send
+        + Sync
+        + 'static,
+    Client::Api: SubspaceApi<Block, FarmerPublicKey> + ObjectsApi<Block>,
+    IQS: ImportQueueService<Block> + ?Sized + 'static,
+{
+    #[allow(clippy::too_many_arguments)]
+    pub(crate) fn new(
+        segment_headers_store: &'a SegmentHeadersStore<AS>,
+        node: &'a Node,
+        piece_getter: &'a PG,
+        client: Arc<Client>,
+        import_queue_service: Arc<Mutex<Box<IQS>>>,
+        network_service: Arc<NetworkService<Block, <Block as BlockT>::Hash>>,
+        fast_sync_state: Arc<parking_lot::Mutex<Option<NumberFor<Block>>>>,
+        sync_service: Arc<SyncingService<Block>>,
+    ) -> Self {
+        Self {
+            segment_headers_store,
+            node,
+            piece_getter,
+            client,
+            import_queue_service,
+            network_service,
+            fast_sync_state,
+            sync_service,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Sync algorithm:
+    /// - 1. download two last segments,
+    /// - 2. add the last block the second last segment (as raw - without checks and execution),
+    /// - 3. download the state for the first block of the last segment,
+    ///     - add the first block of the last segment as raw,
+    ///     - download state for it,
+    /// - 4. add blocks with execution from the last segment,
+    ///     - notify Substrate sync by importing block with using BlockOrigin::NetworkBroadcast,
+    ///     - clear the block gap to prevent Substrate sync to download blocks from the start.
+    ///     - restart the Substrate sync with SyncMode::Full to enable it downloading full blocks.
+    ///  Note: retry for fast sync will degrade reputation if we have already announced blocks.
+    pub(crate) async fn sync(&self) -> Result<FastSyncResult<Block>, Error> {
+        debug!("Starting fast sync...");
+
+        // 1. Download the last two segments.
+        let mut reconstructor = Reconstructor::new().map_err(|error| error.to_string())?;
+        let segment_header_downloader = SegmentHeaderDownloader::new(self.node);
+
+        if let Err(error) = self
+            .download_segment_headers(&segment_header_downloader)
+            .await
+        {
+            error!(?error, "Failed to download segment headers.");
+            return Err(error);
+        };
+
+        let Some(last_segment_index) = self.segment_headers_store.max_segment_index() else {
+            return Err(Error::Other("Can't get last segment index.".into()));
+        };
+
+        // Skip the fast sync if we lack the minimum required segment number
+        if last_segment_index <= SegmentIndex::ONE {
+            return Ok(FastSyncResult::skipped());
+        }
+
+        let last_segment_header = self
+            .segment_headers_store
+            .get_segment_header(last_segment_index)
+            .expect("We get segment index from the same storage. It should be present.");
```

Comment:
This should look more like a proof:
```suggestion
            .expect("Last segment index exists; qed");
```

It is obvious that last segment index returned exists or else it wouldn't be returned as the last once, which is exactly what we needed to prove. Using "shoud" sounds like a wish, not a hard undebatable fact.